src/app/dev/DevToys.Core/Tools/SmartDetectionService.cs

```diff
@@ -0,0 +1,250 @@
+                result.Add(node.Value);
+            }
+            else if (lookup.TryGetValue(node.Value.Detector.Metadata.DataTypeBaseName, out DetectorNode? parentNode))
+            {
+                // If the current detector has a base data type name and its parent node exists in the dictionary,
+                // add it as a child to its parent node
+                parentNode.ChildrenDetectors.Add(node.Value);
+            }
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    /// Builds a dictionary that maps data types to a list of GUI tool instances that accept them.
+    /// </summary>
+    /// <param name="guiToolProvider">The GUI tool provider.</param>
+    /// <returns>A dictionary that maps data types to a list of GUI tool instances that accept them.</returns>
+    private static Dictionary<string, List<GuiToolInstance>> BuildDataTypeToToolInstanceMap(GuiToolProvider guiToolProvider)
+    {
+        // Create a new dictionary with case-insensitive string keys
+        var dataTypeToToolInstanceMap = new Dictionary<string, List<GuiToolInstance>>(StringComparer.OrdinalIgnoreCase);
+
+        // Iterate over all tools provided by the GUI tool provider
+        for (int i = 0; i < guiToolProvider.AllTools.Count; i++)
+        {
+            GuiToolInstance tool = guiToolProvider.AllTools[i];
+
+            // Iterate over all data types accepted by the current tool
+            for (int j = 0; j < tool.AcceptedDataTypeNames.Count; j++)
+            {
+                string dataType = tool.AcceptedDataTypeNames[j];
+                if (!dataTypeToToolInstanceMap.TryGetValue(dataType, out List<GuiToolInstance>? toolList))
+                {
+                    toolList = new();
+                    dataTypeToToolInstanceMap[dataType] = toolList;
+                }
+
+                toolList.Add(tool);
+            }
+        }
+
+        return dataTypeToToolInstanceMap;
+    }
+
+    [LoggerMessage(0, LogLevel.Error, $"Error in {nameof(DetectAsync)}.")]
+    partial void LogDetectAsyncError(Exception ex);
+
+    private sealed class DetectorNode
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DetectorNode"/> class with the specified detector.
+        /// </summary>
+        /// <param name="detector">The detector associated with this node.</param>
+        internal DetectorNode(Lazy<IDataTypeDetector, DataTypeDetectorMetadata> detector)
+        {
+            Guard.IsNotNull(detector);
+            Detector = detector;
+        }
+
+        /// <summary>
+        /// Gets the detector associated with this node.
+        /// </summary>
+        internal Lazy<IDataTypeDetector, DataTypeDetectorMetadata> Detector { get; }
+
+        /// <summary>
+        /// Gets the list of child detectors for this node.
+        /// </summary>
+        internal List<DetectorNode> ChildrenDetectors { get; } = new();
```

Comment:
Some node may not have any child. Let's make this property nullable and only allocate the list if we're actually adding an item?

Reply:
Done